### ISSUE TITLE
[FEAT]  Abstracting the `Model` base class to be re-usable for multiple modelling use cases

### ISSUE THREAD

**Issue by moleary-gsa** [fmind/mlops-python-package#13]
## Description

It would be nice to abstract the [model base class](https://github.com/fmind/mlops-python-package/blob/main/src/bikes/core/models.py#L23-L89) outside the `bikes` package into a more generic package within the repo that could be used for more use cases.

## Motivation

To undertake another modelling task it would seem logical to add another package alongside `bikes`, for example; `taxis` for a NY Taxi data modelling use case. 

A second use case should be able to  make use of most of the base abstractions within the `bikes` package such as the `Job`, `Model`, `Metric` and `Schema` base classes and this is where it feels natural to pull out those abstractions to a package which is generic alongside the `src` package or above it maybe?.

I'm not an expert in python package project layouts but the main point being that the `taxis` and `bikes` packages would be completely separate and depend on a single, framework style package instead of copying across most of the components within the `bikes` package.

## Solutions

I have tried to do this on a fork of the repo [here](https://github.com/m-o-leary/mlops-python-package/tree/mol/refactor-bases-out) but am unable to pass mypy checks because I don't know how to handle the schema dependency on the Model base class. 

In the `bikes` package, the `Model` base class is dependent on the concrete `Inputs`, `Target` and `Outputs` types and I can't seem to figure out how to remove this dependency and avoid the Liskov substitution violation from mypy while keeping the fact that the interface should only accept `pandera` schemas intact.

Any ideas?

**Comment by fmind** [comment]
Hi @moleary-gsa . I had several discussion with my colleague about this idea. It's true there are no "protocol" for MLOps beyond ZenML components, but I'm not sure it would make sense to externalize the abstractions if we don't provide some concretions (e.g., Models + Sklearn + PySparkML + PyTorch, ...).

Moreover, an another argument to keep the model abstraction locally is to use it as a bridge for the project. For instance, the current version has an "get_internal_model" or "explain_samples". By defining them locally, you can fine tune the abstraction to the project need.

My conclusion would be to close the issue, but I'm open to discuss it further if you have counter arguments.
